Objects/obmalloc.c

```diff
@@ -627,6 +661,192 @@ PyObject_SetArenaAllocator(PyObjectArenaAllocator *allocator)
  * around these uses of the runtime-global allocators state. */
 
 
+/************************************/
+/* locking around custom allocators */
+/************************************/
+
+static inline int
+should_lock(PyInterpreterState *interp)
+{
+    return interp->ceval.gil != _PyInterpreterState_Main()->ceval.gil;
+}
+
+static PyThreadState *
+get_tstate_for_main_gil(void)
+{
+    PyThreadState *tstate = _PyRuntimeState_GetFinalizing(&_PyRuntime);
+    if (tstate == NULL) {
+        /* To use its GIL, we only need the pointer and one field. */
+        static const PyThreadState _main_tstate = {
+            .interp = &_PyRuntime._main_interpreter,
+        };
+        tstate = (PyThreadState *)&_main_tstate;
+    }
+    return tstate;
+}
+
+static inline void
+acquire_custom_allocator_lock(PyThreadState *tstate)
+{
+    _PyEval_AcquireLock(tstate);
+}
+
+static inline void
+release_custom_allocator_lock(PyThreadState *tstate)
+{
+    _PyEval_ReleaseLock(tstate->interp, tstate);
+}
+
+static void *
+_PyMem_MallocLocked(void *ctx, size_t size)
+{
+    PyMemAllocatorEx *wrapped = (PyMemAllocatorEx *)ctx;
+    if (_PyRuntime.allocators.num_gils > 1) {
```

Comment:
You lock updates (writes) to this value. Is it safe to use without atomic access or a lock? (requiring an atomic read here would presumably be a performance hit?).

Rather than always checking... could the logic that does `num_gils++` with the (main runtime) lock held also do:

```c
... num_gils++;
if (num_gils == 2 && !has_locking_wrapper(...)) {
    maybe_add_locking_wrapper(...);
}
```

such that the wrapped pointer switch happens upon first creation of an additional gil while the main runtime gil (the _only_ gil prior to the current code running) is still held.

I'm not sure it is worth doing the opposite during finalization. Once wrapped due to per subinterpreter GILs, just stay wrapped. At least the wrappers won't have this conditional anymore.

I suspect this thought is either overoptimization... or actually necessary to avoid locking/atomic access to num_gils.

Reply:
That's a great idea.  I'll try it out.